pkg/controller/ovncontroller/ovncontroller_controller.go

```diff
@@ -0,0 +1,311 @@
+
+	// TemplatesConfigMap
+	templatesConfigMap := ovncontroller.TemplatesConfigMap(instance, instance.Name+"-templates")
+	if err := controllerutil.SetControllerReference(instance, templatesConfigMap, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+	// Check if this TemplatesConfigMap already exists
+	foundTemplatesConfigMap := &corev1.ConfigMap{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: templatesConfigMap.Name, Namespace: templatesConfigMap.Namespace}, foundTemplatesConfigMap)
+	if err != nil && errors.IsNotFound(err) {
+		reqLogger.Info("Creating a new TemplatesConfigMap", "TemplatesConfigMap.Namespace", templatesConfigMap.Namespace, "Job.Name", templatesConfigMap.Name)
+		err = r.client.Create(context.TODO(), templatesConfigMap)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else if !reflect.DeepEqual(templatesConfigMap.Data, foundTemplatesConfigMap.Data) {
+		reqLogger.Info("Updating TemplatesConfigMap")
+		templatesConfigMap.Data = foundTemplatesConfigMap.Data
+	}
+
+	templatesConfigMapHash, err := util.ObjectHash(templatesConfigMap.Data)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error calculating configuration hash: %v", err)
+	}
+	reqLogger.Info("TemplatesConfigMapHash: ", "Data Hash:", templatesConfigMapHash)
+
+	// Define a new Daemonset object
+	ds := newDaemonset(instance, instance.Name, templatesConfigMapHash)
+	dsHash, err := util.ObjectHash(ds)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error calculating configuration hash: %v", err)
+	}
+	reqLogger.Info("DaemonsetHash: ", "Daemonset Hash:", dsHash)
+
+	// Set OVNController instance as the owner and controller
+	if err := controllerutil.SetControllerReference(instance, ds, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if this Daemonset already exists
+	found := &appsv1.DaemonSet{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: ds.Name, Namespace: ds.Namespace}, found)
+	if err != nil && errors.IsNotFound(err) {
+		reqLogger.Info("Creating a new Daemonset", "Ds.Namespace", ds.Namespace, "Ds.Name", ds.Name)
+		err = r.client.Create(context.TODO(), ds)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Daemonset created successfully - don't requeue
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{}, err
+	} else {
+		if instance.Status.DaemonsetHash != dsHash {
+			reqLogger.Info("Daemonset Updated")
+			found.Spec = ds.Spec
+			err = r.client.Update(context.TODO(), found)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			r.setDaemonsetHash(instance, dsHash)
+			return reconcile.Result{RequeueAfter: time.Second * 10}, err
+		}
+	}
+
+	// Daemonset already exists - don't requeue
+	reqLogger.Info("Skip reconcile: Daemonset already exists", "Ds.Namespace", found.Namespace, "Ds.Name", found.Name)
+	return reconcile.Result{}, nil
```

Comment:
perhaps we want to add a timing reconciliation here (similarly to the computenode operator), to ensure every X mins (5-10 mins) that everything needed is still there

Reply:
The compute-node-operator pattern is different though. If we add an explicit c.Watch on Daemonsets that would immediately trigger the operator to run if there are any changes (see review comment above). Periodically re-running the operator is good too but lets not rely on that for Daemonset change detection as those are a key component here.